File: Data-Structures-and-Algorithms-Specialization/1. Algorithmic Toolbox/Week 2/5_fibonacci_number_again/fibonacci_huge.cpp
#include <iostream>
using namespace std;

int pisano_period(int m) {
    int previous = 0;
    int current = 1;
    for (int i = 0; i < m * m; i++){
        previous = current;
        current = (previous + current) % m;
        if (previous == 0 && current == 1)
            return  i + 1;
    }
    return 1;
}

int fibonacci_huge(int n, int m) {
    int pp_index = pisano_period(m);
    n = n % pp_index;
    int previous = 0;
    int current = 1;
    for (int i = 0; i < n-1; i++){
        previous = current;
        current = previous + current;
    }
    return current % m;
}
    

int main(){
    int m, n;
    cin >> n, m;
    cout << fibonacci_huge(n, m);
    return 0;
}

File: Data-Structures-and-Algorithms-Specialization/1. Algorithmic Toolbox/Week 2/5_fibonacci_number_again/fibonacci_huge.py
def pisano_period(m):
    previous, current = 0, 1
    for i in range(0, m * m):
        previous, current = current, (previous + current) % m
        if previous == 0 and current == 1:
            return i + 1


def fibonacci_huge(n, m):
    pp_index = pisano_period(m)
    n = n % pp_index
    previous, current = 0, 1
    if n <= 1:
        return n
    for i in range(n - 1):
        previous, current = current, previous + current
    return current % m


if __name__ == '__main__':
    n, m = map(int, input().split())
    print(fibonacci_huge(n, m))
